elephant/spectral.py

```diff
@@ -242,6 +242,121 @@ def welch_psd(signal, n_segments=8, len_segment=None,
+    using Multitaper method
+
+    The PSD is obtained through the following steps:
+
+    1. Calculate 'num_tapers' approximately independent estimates of the
+       spectrum by multiplying the signal with the discrete prolate spheroidal
+       functions (also known as Slepian function) and calculate the PSD of the
+       products
+
+    2. Average the approximately independent estimates to decrease overall
+       variance of the estimates
+
+    Parameters
+    ----------
+    signal : neo.AnalogSignal
+        Time series data of which PSD is estimated. When `signal` is np.ndarray
+        sampling frequency should be given through keyword argument `fs`.
+    fs : float, optional
+        Specifies the sampling frequency of the input time series
+        Default: 1.0
+    nw : float, optional
+        Time bandwidth product
+        Default: 4.0
+    num_tapers : int, optional
+        Number of tapers used in 1. to obtain estimate of PSD. By default
+        [2*nw] - 1 is chosen.
+        Default: None
+    peak_resolution : float, optional
+        Desired frequency resolution of the obtained PSD estimate. When given
+        as a `float`, it is taken as frequency in Hz.
+        If None, it will be determined from other parameters.
+        Default: None.
+
+    Returns
+    -------
+    freqs : np.ndarray
+        Frequencies associated with power estimate in `psd`
+    psd : np.ndarray
+        PSD estimate of the time series in `signal`
+    """
+
+    # When the input is AnalogSignal, the data is added after rolling the axis
+    # for time index to the last
+    data = np.asarray(signal)
+    if isinstance(signal, neo.AnalogSignal):
+        data = np.rollaxis(data, 0, len(data.shape))
+
+    # If the data is given as AnalogSignal, use its attribute to specify the
+    # sampling frequency
+    if hasattr(signal, 'sampling_rate'):
+        fs = signal.sampling_rate.rescale('Hz').magnitude
+
+    # If fs and frequency resolution is pq Quantity get magnitude
+    if isinstance(fs, pq.quantity.Quantity):
+        fs = fs.rescale('Hz').magnitude
+    if isinstance(peak_resolution, pq.quantity.Quantity):
+        peak_resolution = peak_resolution.rescale('Hz').magnitude
+
+    # Number of data points in time series
+    if data.ndim == 1:
+        length_signal = np.shape(data)[0]
+    else:
+        length_signal = np.shape(data)[1]
+
+    # Determine time-halfbandwidth product from given parameters
+    if peak_resolution is not None:
+        if peak_resolution <= 0:
+            raise ValueError("peak_resolution must be positive")
+        else:
+            nw = length_signal / fs * peak_resolution / 2
+
+    if num_tapers is None:
+        num_tapers = np.floor(2*nw).astype(int) - 1
+    else:
+        if not isinstance(num_tapers, int):
+            raise TypeError("num_tapers must be integer")
+        elif num_tapers <= 0:
+            raise ValueError("num_tapers must be positive")
```

Comment:
Check if the optional parameters are `None`.

Reply:
We believe that this is not necessary, because if `peak_resolution` is `None`, it does not affect the params. If `num_tapers` is `None`, the `num_tapers` is computed from `nw`, which has a default value of 4.